src/modules/m_clones.cpp

```diff
@@ -21,19 +21,112 @@
 
 #include "inspircd.h"
 
+typedef std::vector<User*> UList;
+typedef std::map<irc::sockets::cidr_mask, UList> CloneList;
+
+class UserTracker : public Timer
+{
+	bool tracking;
+	unsigned long timeout;
+	CloneList cl;
+
+public:
+	UserTracker()
+		: Timer(timeout, false), tracking(false), timeout(0)
+	{
+	}
+
+	bool Tick(time_t) CXX11_OVERRIDE
+	{
+		StopTracking();
+		return false;
+	}
+
+	void SetTimeout(unsigned long t)
+	{
+		timeout = t;
+	}
+
+	const CloneList& GetMap()
+	{
+		if (tracking)
+		{
+			SetInterval(timeout);
+			return cl;
+		}
+
+		if (timeout)
+		{
+			tracking = true;
+			SetInterval(timeout);
+		}
+
+		for (user_hash::const_iterator it = ServerInstance->Users->clientlist.begin(); it != ServerInstance->Users->clientlist.end(); ++it)
+		{
+			User* user = it->second;
+			AddClient(user);
+		}
+		return cl;
+	}
+
+	void StopTracking()
+	{
+		cl.clear();
+		tracking = false;
+	}
+
+	void AddClient(User* user)
+	{
+		if (!tracking)
+			return;
+
+		UList& counts = cl[user->GetCIDRMask()];
+		counts.push_back(user);
+	}
+
+	void DelClient(User* user)
+	{
+		if (!tracking)
+			return;
+
+		CloneList::iterator i = cl.find(user->GetCIDRMask());
+		if (i == cl.end())
+		{
+			ServerInstance->Logs->Log(MODNAME, LOG_DEFAULT, "Potential bug. CIDR was not found in clone map upon removal. Did you change clone CIDR?");
+			return StopTracking();
+		}
+
+		UList& counts = i->second;
+		UList::iterator it = std::find(counts.begin(), counts.end(), user);
+		if (it == counts.end())
+		{
+			ServerInstance->Logs->Log(MODNAME, LOG_DEFAULT, "Potential bug. User was not found in clone map upon removal. Did you change clone CIDR?");
+			return StopTracking();
+		}
+
+		stdalgo::vector::swaperase(counts, it);
```

Comment:
The above if should check the return of this too, if it returns false the user was not found and the same should happen